src/gpio.rs

```diff
@@ -1107,8 +1107,24 @@ where
     }
 }
 
+#[cfg(not(feature = "riscv-ulp-hal"))]
+use crate::embedded_hal_error;
+#[cfg(not(feature = "riscv-ulp-hal"))]
+embedded_hal_error!(
+    GpioError,
+    embedded_hal::digital::Error,
+    embedded_hal::digital::ErrorKind
+);
+#[cfg(not(feature = "riscv-ulp-hal"))]
+fn to_gpio_err(err: EspError) -> GpioError {
+    GpioError::other(err)
+}
+
 impl<'d, T: Pin, MODE> embedded_hal::digital::ErrorType for PinDriver<'d, T, MODE> {
```

Comment:
this is kinda hacky... the embedded_hal_error! doesnt work for the ulp_processor .... so a lot of feature flags. Ideas how to do that better welcome.

Reply:
Maybe try to make `embedded_hal_error` compatible with the ulp mini-HAL? `EspError` [does exist when the mini-HAL is enabled](https://github.com/esp-rs/esp-idf-hal/blob/master/src/riscv_ulp_hal/sys.rs#L10), so perhaps you can conditionalize the macro body itself, rather than conditionalizing all places where it is used?

Reply:
did it in fe4875cef944188aa6b6b44936fa09b4cf0b73c7 though i have a follow up question, it builds and seem to work, but i thought that one should use absolute path, inside a macro, though it works if i use the cfg feature like that outside the macro. What is your opinion on that? This way the macro is cleaner to read but i am not sure if this is correct.

Reply:
i duplicate the macro now if used in ulp context d877faa to circumvent possible macro errors